### PR TITLE
[Example] Training offloading example

### PR DESCRIPTION
**Pull request by songgot**
- Add nnstreamer example for training offloading
- Add README.md

### CODE REVIEW

File: native/example_training_offloading/meson.build [hunk 1]
```diff
@@ -0,0 +1,6 @@
+nnstreamer_example_training_offloading = executable('nnstreamer_example_training_offloading',
+  'nnstreamer_example_training_offloading.c',
+  dependencies: [glib_dep, gst_dep, nns_dep, nns_edge_dep],
```
[the remark this answers is not in the document]
Comment (songgot) [reply native/example_training_offloading/meson.build:3]: Thank you, I will add dependency.

**Comment by songgot** [comment]
Thank you for your review @gichan-jang

File: native/example_training_offloading/README.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+![receiver.png](./receiver.png)
+
+### How to Run
+Run the mosquitto service and check that the status is Active .
```
Comment (jaeyun-jung) [inline native/example_training_offloading/README.md:20]: remove space: `Active .` > `Active.`

Comment (songgot) [reply native/example_training_offloading/README.md:20]: Thank you for your review!

File: native/example_training_offloading/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+---
+title: Training offloading example
+...
+
+## Ubuntu Native NNStreamer Application Example - Training offloading
+### Introduction
+This example implements training offload between two devices in the nnstreamer-edge(https://github.com/nnstreamer/nnstreamer-edge) network environment. this operates as a sender or a receiver according to the stream-role value input from the user. the sender reads the data required for training from ```datareposrc``` and uses ```edgesink``` to send the data to the peer device on the network. The receiver reads the received data from the peer device using ```edgesrc``` and pushes it to tensor_trainer. then tensor_trainer trains the model using the data. data transmission and reception between ```edgesink``` and ```edgesrc``` in the network is carried out through MQTT broker. user does not need to know the details and only needs to match the ```dest-host(IP)```, ```dest-port```, ```connect-type``` and ```topic``` of the MQTT broker to connect ```edgesrc``` and ```edgesink```.
```
Comment (jaeyun-jung) [inline native/example_training_offloading/README.md:7]: Let's fix typo and link in readme.

This example implements training offload between two devices in the [nnstreamer-edge](https://github.com/nnstreamer/nnstreamer-edge) network environment. This operates as a sender or a receiver according to the stream-role value input from the user. The sender reads the data required for training from ```datareposrc``` and uses ```edgesink``` to send the data to the peer device on the network. The receiver reads the received data from the peer device using ```edgesrc``` and pushes it to tensor_trainer. Then tensor_trainer trains the model using the data. Data transmission and reception between ```edgesink``` and ```edgesrc``` in the network is carried out through MQTT broker. User does not need to know the details and only needs to match the ```dest-host(IP)```, ```dest-port```, ```connect-type```, and ```topic``` of the MQTT broker to connect ```edgesrc``` and ```edgesink```.

File: native/example_training_offloading/README.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+#### sender pipeline ####
+The sender is configured as follows. ```datareposrc``` read data from the location path with ```JSON``` meta information. ```datareposrc``` repeatedly reads the data as many as ```epochs``` from ```start-sample-index``` to ```stop-sample-index``` among the entire data. the default value of ```epochs``` is 1, the default value of ```start-sample-index``` is 0 and the default value of ```stop-sample-index``` is the total number of samples -1. If you use the default value, you do not need to set the value. ```edgesink``` uses nnstreamer-edge network environment. ```edgesink``` connect ```MQTT broker``` so it needs to run ```systemctl start mosquitto```. the connection type of ```edgesink``` is ```HYBRID``` and the topic is ```tempTopic```.
+
+<img src="./sender.png" width="50%" heigth="50%"/>
+
+#### sender pipeline ####
```
Comment (jaeyun-jung) [inline native/example_training_offloading/README.md:14]: receiver pipeline maybe

File: native/example_training_offloading/README.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+The sender is configured as follows. ```datareposrc``` read data from the location path with ```JSON``` meta information. ```datareposrc``` repeatedly reads the data as many as ```epochs``` from ```start-sample-index``` to ```stop-sample-index``` among the entire data. the default value of ```epochs``` is 1, the default value of ```start-sample-index``` is 0 and the default value of ```stop-sample-index``` is the total number of samples -1. If you use the default value, you do not need to set the value. ```edgesink``` uses nnstreamer-edge network environment. ```edgesink``` connect ```MQTT broker``` so it needs to run ```systemctl start mosquitto```. the connection type of ```edgesink``` is ```HYBRID``` and the topic is ```tempTopic```.
```
Comment (jaeyun-jung) [inline native/example_training_offloading/README.md:10]: The sender is configured as follows. ```datareposrc``` reads data from the location path with ```JSON``` meta information. ```datareposrc``` repeatedly reads the data as many as ```epochs``` from ```start-sample-index``` to ```stop-sample-index``` among the entire data set. The default value of ```epochs``` is 1, the default value of ```start-sample-index``` is 0 and the default value of ```stop-sample-index``` is the total number of samples -1. If you use the default value, you do not need to set the value. ```edgesink``` uses nnstreamer-edge network environment. ```edgesink``` connects to ```MQTT broker``` so it needs to run ```systemctl start mosquitto```. The connection type of ```edgesink``` is ```HYBRID``` and the topic is ```tempTopic```.

File: native/example_training_offloading/README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+<img src="./sender.png" width="50%" heigth="50%"/>
+
+#### sender pipeline ####
+The receiver is configured as follows. to receive data from a peer device, ```edgesrc``` uses nnstreamer-edge network environment. as with the sender, ```edgesrc``` connect MQTT broker so it needs to run ```systemctl start mosquitto```. the connection type of ```edgesink``` is ```HYBRID``` and the topic is ```tempTopic```. now, the received data is passed to tensor_trainer. set the ```framework``` to use for training the model, configure the model with ```model-config``` file and set ```model-save-path``` to save a model. for input caps of tensor_trainer, refer to gst_caps in JSON file or check the input format of model-config. users will know the format of the data used for model training and the number of inputs and labels. preprocessed data affects the performance of model training. set ```num-inputs``` and ```num-labels```(both default value is 1). it needs to set how many of the input data to use for training and validation for model training and set the number of epochs. the properties used for set are ```num-training-samples```, ```num-validation-samples```, and ```epochs``` respectively.
```
Comment (jaeyun-jung) [inline native/example_training_offloading/README.md:15]: The receiver is configured as follows. To receive data from a peer device, ```edgesrc``` uses nnstreamer-edge network environment. As with the receiver, ```edgesrc``` connects to MQTT broker so it needs to run ```systemctl start mosquitto```. The connection type of ```edgesrc``` is ```HYBRID``` and the topic is ```tempTopic```. Now, the received data is passed to tensor_trainer. Set the ```framework``` to use for training the model, configure the model with ```model-config``` file and set ```model-save-path``` to save a model. For input caps of tensor_trainer, refer to gst_caps in JSON file or check the input format of model-config. Users will know the format of the data used for model training and the number of inputs and labels. The preprocessed data affects the performance of model training. Set ```num-inputs``` and ```num-labels```(both default value is 1). It needs to set how many of the input data being used for training and validation for model training, and set the number of epochs. The properties for these are ```num-training-samples```, ```num-validation-samples```, and ```epochs``` respectively.

File: native/example_training_offloading/README.md (hunk 3 above)
Comment (songgot) [reply native/example_training_offloading/README.md:7]: Thank you.

File: native/example_training_offloading/README.md (hunk 4 above)
Comment (songgot) [reply native/example_training_offloading/README.md:14]: sender is right. edgesink is in the sender pipeline.

File: native/example_training_offloading/README.md (hunk 5 above)
Comment (songgot) [reply native/example_training_offloading/README.md:10]: Thank you.

File: native/example_training_offloading/README.md (hunk 6 above)
Comment (songgot) [reply native/example_training_offloading/README.md:15]: Thank you.

File: native/example_training_offloading/README.md (hunk 4 above)
Comment (songgot) [reply native/example_training_offloading/README.md:14]: Thank you.

**Comment by songgot** [comment]
Thank you for your detailed review.

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**